src/Logic.hs

```diff
@@ -87,154 +84,124 @@ import qualified GithubApi
 import qualified Project as Pr
 import qualified Time
 
--- | There are two categories of actions, semantically distinct. BaseActionFree will
--- actually be translated into actions, whereas RetrieveEnvironment is more focused
--- on getting values from the context we're in. Some things could still be moved around
--- though.
-data BaseActionFree a
-  = TryIntegrate
+-- | Represents the high level manipulations of a pull request
+data Action :: Effect where
     -- This is a record type, but the names are currently only used for documentation.
+  TryIntegrate ::
     { _mergeCommitMessage   :: Text
     , _integrationCandidate :: (PullRequestId, Branch, Sha)
     , _train                :: [PullRequestId]
     , _alwaysAddMergeCommit :: Bool
-    , _cont                 :: Either IntegrationFailure Sha -> a
-    }
-  | TryPromote Branch Sha (PushResult -> a)
-  | TryPromoteWithTag Branch Sha TagName TagMessage (PushWithTagResult -> a)
-  | CleanupTestBranch PullRequestId a
-  | LeaveComment PullRequestId Text a
-  | IsReviewer Username (Bool -> a)
-  | GetPullRequest PullRequestId (Maybe GithubApi.PullRequest -> a)
-  | GetOpenPullRequests (Maybe IntSet -> a)
-  | GetLatestVersion Sha (Either TagName Integer -> a)
-  | GetChangelog TagName Sha (Maybe Text -> a)
-  | IncreaseMergeMetric a
-  | UpdateTrainSizeMetric Int a
-  deriving (Functor)
-
-data RetrieveEnvironmentFree a
-  = GetProjectConfig (ProjectConfiguration -> a)
-  | GetDateTime (UTCTime -> a)
-  | GetBaseBranch (BaseBranch -> a)
-  deriving (Functor)
-
-type BaseAction = Free BaseActionFree
-type RetrieveEnvironment = Free RetrieveEnvironmentFree
-
--- | The actions represented here will be tranformed into a lower-level type of actions,
--- embodied in the Operation monad.
-type Action = Free (Sum BaseActionFree RetrieveEnvironmentFree)
-
-type Operation = Free (Sum
-                        (Sum
-                          MetricsOperationFree
-                          TimeOperationFree)
-                        (Sum
-                          GitOperationFree
-                          GithubOperationFree))
```

Comment:
Farewell, my tree-shaped friend. We shall mourn your loss, but rejoice in the newly discovered pleasures of `effectful`.